### ISSUE TITLE
Burn does not pass restart flags to ExePackage bundles

### ISSUE THREAD

**Issue by nirbar** [wixtoolset/issues#6252]
* Which version of WiX are you building with?
3.11.2.4516

* Which version of Visual Studio are you building with (if any)?
VS 2019.6.3

* Which version of the WiX Toolset Visual Studio Extension are you building with (if any)?
3.11.2.4516

* Which version of .NET are you building with?
4.5

* If the problem occurs when installing your packages built with WiX, what is the version of Windows the package is running on?
Windows Server 2016

* Describe the problem and the steps to reproduce it.
  1. Create a bootstrapper with version 1.0.0
  1. Rebuild the same bootstrapper with version 2.0.0
  1. Install bootstrapper v1.0.0
  1. Install bootstrapper v2.0.0 and specify "/norestart" on the command line

* Describe the behavior you expected and how it differed from the actual behavior.
  - Expected behavior:
    - Bootstrapper v2.0.0 will perform an uninstall of v1.0.0 and specify "/norestart" on the process command arguments
  - Actual behavior:
    - Bootstrapper v2.0.0 performs an uninstall of v1.0.0 and doesn't specify "/norestart" on the process command arguments. The related bundle v1.0.0 uses 'automatic' as restart option

I'd expect the related bundle reboot option to be dependent upon the parent bundle's option as following:
  - Parent "norestart" --> related "norestart"
  - Parent "forcerestart" --> related "norestart" (parent will reboot on end)
  - Parent "promptrestart" --> related "norestart" (parent will prompt on end)
  - Parent automatics restart --> related "norestart" (parent will reboot if needed on end)

Bottom line- related bundles should always be executed with "norestart" argument

**Comment by rseanhall** [comment]
> Bottom line- related bundles should always be executed with "norestart" argument

I don't know about this, need to think about the scenario where the related bundle stopped mid-chain because of a reboot. The parent bundle would also need to stop mid-chain and execute the related bundle again after the reboot.

**Comment by nirbar** [comment]
The fix would prevent the related bundle from rebooting mid-chain

**Comment by rseanhall** [comment]
No, a bundle needs to be able to stop mid-chain because one of its packages requires a reboot. Packages later on might not execute properly without the reboot.

**Comment by nirbar** [comment]
@rseanhall If the related bundle has to reboot midway then it should exit with exit code requiring reboot and delegate the decision to the parent.
The parent bundle would then act based on it's command line reboot option.

After the reboot, the related bundle would resume automatically per its registration resume command.
The parent bundle may as well resume after the reboot if it needs to

**Comment by barnson** [comment]
It should work but we need to consider whether it affects other cases. The behavior change means it's something to consider in v4.

**Comment by nirbar** [comment]
I think that the fix needs to attend the following cases as well:
- Parent bundle is with an older engine version, so any related bundle should default to 'norestart'
- ExePackage with burn protocol - both when the parent bundle has an older engine version or the ExePackage has an older engine version

**Comment by nirbar** [comment]
I think that the fix needs to attend the following cases as well:
- Parent bundle is with an older engine version, so any related bundle should default to 'norestart'
- ExePackage with burn protocol - both when the parent bundle has an older engine version or the ExePackage has an older engine version

**Comment by rseanhall** [comment]
[This code](https://github.com/wixtoolset/codeplex/commit/3c56c1d41eff0b1fe6cc04fcdd742ced7f554ca4) fixed [this issue](https://sourceforge.net/p/wix/bugs/3109/) back in 3.7. Can you please explain how that's failing to disable related bundles from restarting? What's the actual problem you're running into?

```
// If this is a related bundle (but not an update) suppress restart and return the standard restart error code.
    if (fRestart && BOOTSTRAPPER_RELATION_NONE != engineState.command.relationType && BOOTSTRAPPER_RELATION_UPDATE != engineState.command.relationType)
    {
        LogId(REPORT_STANDARD, MSG_RESTART_ABORTED, LoggingRelationTypeToString(engineState.command.relationType));

        fRestart = FALSE;
        hr = HRESULT_FROM_WIN32(ERROR_SUCCESS_REBOOT_REQUIRED);
    }
```

**Comment by nirbar** [comment]
I see what you mean
I think that it doesn't handle ExePackage with burn protocol does it?

**Comment by rseanhall** [comment]
My vote is that this should only be a feature for [`BundlePackage`](https://github.com/wixtoolset/issues/issues/3693). `Protocol="burn"` is only intended to be for progress.

**Comment by barnson** [comment]
Burn protocol shouldn't be overloaded like that.

**Comment by rseanhall** [comment]
To support a bundle being included in the chain as an ExePackage (like the .NET Core redistributables), I'm going to add `ExePackage/@WixBundle=yes` so the user can opt-in to the v3 behavior. Burn will add the restart flags to those ExePackages.

**Comment by barnson** [comment]
`WixBundle`? Is `Bundle` sufficient?

**Comment by rseanhall** [comment]
Is Bundle really that specific to Wix?

**Comment by barnson** [comment]
In WiX authoring, yes. :)
